python/cuml/preprocessing/TargetEncoder.py

```diff
@@ -0,0 +1,213 @@
+#
+# Copyright (c) 2019, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import cudf as gd
+import cupy as cp
+from cuml.common.exceptions import NotFittedError
+
+
+class TargetEncoder:
+    """
+    A cudf based implementation of target encoding
+
+    Parameters
+    ----------
+    n_folds : int (default=5)
+        Default number of folds
+    smooth : int (default=0)
+        Number of samples to smooth the encoding
+    seed : int (default=42)
+        Random seed
+    split : {'random', 'continuous', 'interleaved'}, default='interleaved'
+    Examples
+    --------
+    Converting a categorical implementation to a numerical one
+
+    .. code-block:: python
+
+        from cudf import DataFrame, Series
+
+        train = DataFrame({'category': ['a', 'b', 'b', 'a'],
+                           'label': [1, 0, 1, 1]})
+        test = DataFrame({'category': ['a', 'c', 'b', 'a']})
+
+        # There is only one correct way to do this
```

Comment:
This is true but could maybe be expanded to like: "You must use fit_transform on train then transform on test data"

Reply:
Does this work on one categorical variable (or set of columns treated as one category) at a time? Or several completely independent ones? Is there any serious perf advantage to batching over several?

Reply:
Yes, it works on one categorical variable at a time. However, encoding multiple columns in one `fit` is not for performance but for joint encoding. For example,

```
# encode each column independently and one output for each input column.
for column in ['age', 'gender']:
    train[column+'_TE'] = encoder.fit_transform(train[column], train['income'])
   # is similar to:
   # train[column+'_TE'] = train.groupby(column).transform({'income': 'mean'})
```


```
# encode both columns jointly and only one output for all columns.
train['age_gender_TE'] = encoder.fit_transform(train[['age', 'gender']], train['income'])
# is similar to
# train[age_gender_TE'] = train.groupby(['age', 'gender']).transform({'income': 'mean'})
```